File: leetcode_0078/cpp/leetcode_0078.cpp
//
// Created by wzy on 2022/4/18.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;

class Solution {
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        //思路：回溯
        vector<vector<int>> res;
        int sz = nums.size();
        vector<int> sub;
        //依次设置子集的元素个数
        for(int i = 0;i <= sz;i++)
            // sub.clear();
            dfs(res,sub,nums,0,i);
        return res;
    }
private:
    void dfs(vector<vector<int>>& res,vector<int>& sub,vector<int>& nums,int start,int k){
        if(sub.size() == k){
            res.push_back(sub);
            return ;
        }
        for(int i = start;i < nums.size();i++){
            //如果后序元素加上已有元素不足k则直接返回
            if(sub.size() + nums.size() - i + 1 < k)
                return ;
            sub.push_back(nums[i]);
            dfs(res,sub,nums,i+1,k);
            sub.pop_back();
        }
    }
};


template<typename T>
string convertVecToStr(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    return res;
}

template<typename T>
void printVecArr(std::vector<std::vector<T>> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << convertVecToStr(ele);
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main(){
    vector<int> nums = {1,2,3};
    //nums = [1,2,3]
    auto res = Solution().subsets(nums);
    printVecArr(res);
    //[],[1],[2],[3],[1,2],[1,3],[2,3],[1,2,3]]
}

File: leetcode_0078/java/leetcode_0078.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class leetcode_0078 {
    class Solution {
        public List<List<Integer>> subsets(int[] nums) {
            //思路：回溯
            int sz = nums.length;
            List<List<Integer>> res = new LinkedList();
            LinkedList<Integer> sub = new LinkedList();
            for(int i = 0;i <= sz;i++)
                dfs(res,sub,nums,0,i);
            return res;
        }
        private void dfs(List<List<Integer>> res,LinkedList<Integer> sub,int[] nums,int start,int k){
            if(sub.size() == k){
                res.add(new LinkedList(sub));
                return ;
            }
            for(int i = start;i < nums.length;i++){
                //如果后序元素加上已有元素不足k则直接返回
                if(sub.size() + nums.length - i + 1 < k)
                    return ;
                sub.addLast(Integer.valueOf(nums[i]));
                dfs(res,sub,nums,i+1,k);
                sub.removeLast();
            }
        }
        public <T> void printListArr(List<List<T>> list) {
            for (int i = 0; i < list.size(); i++) {
                if (i == list.size() - 1) {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.println(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " ] ");
                } else {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.printf(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " , ");
                }
            }
        }
    }
    public static void main(String[] args) {
        leetcode_0078.Solution solution = new leetcode_0078().new Solution();
        int[] nums = {1,2,3};
        List<List<Integer>> res = solution.subsets(nums);
        solution.printListArr(res);
        //[ [] , [1] , [2] , [3] , [1, 2] , [1, 3] , [2, 3] , [1, 2, 3] ]
    }
}
